Dictionary of Combinations

jduong09 opened the pull request:
# Feature

## Dictionary Function

### Summary
- Created dictionary of combinations.
- Modified validateCombo function in Game Component to test combos based on current turn's given combination.

### How To Test
- Through localhost:3000
   - Run server through `npm run dev`
   - Shuffle deck and browser will display hand of 13 cards and prompt user to finalize a turn that matches the combo prompt.
  
- Through node.js
   - `node components/utilities/combination.js`
   - Validate terminal log

jduong09 commented:
## Update

- Changed numbers array when building card deck, changing J, Q, K, A, 2 to equal their number value; 11, 12, 13, 14, 15.

ToDo: Fix display on browser to match number value to their correct card display.

ziggysauce commented on components/game/game.js:
```diff
@@ -1,17 +1,20 @@
   // Build Card Deck
   const suites = ['spades', 'clubs', 'diamonds', 'hearts'];
-  const numbers = ['3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A', '2'];
+  const numbers = ['3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15'];
```
this is probably fine actually. i would just leave a comment here somewhere indicating what 11-15 represent. as long as in our UI we properly display them as Jack, Queen, King, Ace, and 2 we are all good with this.

ziggysauce commented on components/utilities/combination.js:
```diff
@@ -0,0 +1,176 @@
+        return false;
+      }
+    
+      return cards[0].number === cards[1].number && cards[0].number === cards[2].number && cards[0].number === cards[3].number;
```
this will be the same for pair, triplet and quartet, but you can optimize it by just grabbing the first card number, then making sure each card after is the same

```javascript
const [{number: cardNumber}] = cards; // gets the first card's number
cards.every(({number}) => number === cardNumber);
```

or something to this effect. if you really wanna make it concise, something like this would work:

```javascript
isValid: (cards) => {
  const [{number: cardNumber}] = cards; // gets the first card's number
  return cards.length === 4 && cards.every(({number}) => number === cardNumber);
}
```

^^ this checks that both the length is 4 AND each card is the same in nice lil concise way.

jduong09 replied:
This is beautiful, thanks.

ziggysauce commented on components/utilities/combination.js:
```diff
@@ -0,0 +1,176 @@
+        return false;
+      }
+    
+      let containsDupe = false;
+      const uniqNumbers = {};
+      for (let i = 0; i < cards.length; i++) {
+        if (uniqNumbers[cards[i].number] === undefined) {
+          uniqNumbers[cards[i].number] = true;
+        } else {
+          containsDupe = true;
+        }
```
can consolidate the logic using a reduce like so (this is untested, so pls test):

```javascript
const {uniqNumbers, containsDupe} = cards.reduce((acc, {number}) => {
  if(acc.uniqNumbers.includes(number)) {
    acc.containsDupe = true;
  } else {
    acc.uniqNumbers.push(number);
  }
  return acc;
}, {uniqNumbers: [], containsDupe: false});
```

this will give you the same boolean value for `containsDupe` as well as an array of the numbers so you don't need to re-iterate through the cards array and create an array from an object like you are below.

jduong09 replied:
This is great refactoring, I used this template for double sequence as well.

ziggysauce reviewed (changes requested):
just some preliminary review to optimize some logic

ziggysauce commented on components/game/game.js:
```diff
@@ -61,6 +64,7 @@ const Game = () => {
     setTimeout(() => {
       showIntro(false);
       shuffleDeck(true);
+      comboChoiceLoop()
```
```suggestion
      comboChoiceLoop();
```

ziggysauce commented on components/game/game.js:
```diff
@@ -103,6 +110,43 @@ const Game = () => {
+   * @description Simulates user selecting combos.
+   */
+  const comboChoiceLoop = () => {
+    const randomInt = Math.floor(Math.random() * (6 - 1 + 1) + 1);
```
as discussed, remove this `randomInt` logic and just show all the combo choice options with the ability to choose one. you'll need to add that to state. we can either comment this all out or fully remove it later after testing is done.

jduong09 commented:
## Update

- Refactored logic on combination.js
- Added helper function to display face card based on number in cards array
- Removed logic to randomize a combination for testing, added a select tag that allows user to manually choose which combo they want to test

ziggysauce commented on components/game/game.js:
```diff
@@ -1,17 +1,21 @@
 import gameStyles from './game.module.scss';
 import { useState } from 'react';
 import Hand from "@/components/gameComponents/hand.js";
+import { dictionaryCombinations } from '@/components/utilities/combination.js';
```
what's with the `@/components/...` nomenclature? i've never seen the `@` before a path, is this something specific to react/next?

jduong09 replied:
https://nextjs.org/docs/app/building-your-application/configuring/absolute-imports-and-module-aliases

- Next.js has built in way to handle relative paths

ziggysauce replied:
ahhh nifty!

ziggysauce commented on components/game/game.js:
```diff
@@ -118,14 +151,31 @@ const Game = () => {
       {deckIsShuffled &&
         <div>
           <h2 className={gameStyles.turnIndicator}>{playerTurn === 0 ? 'Your' : `Player ${playerTurn + 1}'s`} turn.</h2>
+          <h2>{`Select a combo thats fits ${selectCombo}`}</h2>
```
the whole thing doesn't need to be in string literals. in fact, i think only the variable is needed, no string literals at all

```suggestion
          <h2>Select a combo thats fits {selectCombo}</h2>
```

ziggysauce reviewed (changes requested):
nice! this tested pretty well for me. just a few things otherwise this is probs good to go. i'd get rid of most/all the console logs before merging too.

---

hmm, i don't see the helper function to display the face cards. my UI still looks like this

<img width="917" alt="Screen Shot 2024-08-19 at 8 38 53 PM" src="https://github.com/user-attachments/assets/b0bdf81d-31d3-4dd2-964b-0abb2e511a0b">

jduong09 commented:
## Comment on Review

- SORRY, PR that had the helper function was not pushed to github. Pull down the helper function commit, and the changes I made this morning and it should be updated.

ziggysauce reviewed (approved):
nice! this looks good to me. the only thing i think we can re-visit is the `numbers` array. now that we're using values 3-15, they can probably just be stored as numbers (ints) instead of strings so that doing the mapping in your `mapCard` is slightly faster

jduong09 commented:
## Update 

Merged PR.
